### DISCUSSION TITLE
提高音韻地位分辨能力（及相應的增強）

### DISCUSSION THREAD

**Discussion by syimyuzya** [nk2028/qieyun-js#25] [Ideas]
tl;dr 音韻地位與資料脫鉤；放寬音韻地位檢查；推導方案API增強；提升自帶音韻資料的質量

## 為什麼要做這個？

前天（2022-03-24）在和 @ayaka14732 討論中再次提到了目前的音韻地位體系會將過多小韻合併不分的問題。

如今切韻拼音已經公開，輸入法碼表也在計劃在上遊以音韻地位編排，但目前的音韻地位區分能力既不夠用於切韻拼音，也不夠用於碼表，甚至不夠用於《韻典》。因此，增加音韻地位的對《廣韻》小韻的區分能力，以及增加區分《切韻》諸版本、切韻系韻圖中的一些額外對立這兩點，**不久將成為必然需求**。

於是想了一天解決方案，從保守方式到激進方式都考慮了下，這裡先放一下目前算是比較成形的想法。

### 問題一：音韻地位無法表述切韻拼音、韻典等系統的音韻地位	

這個問題前天已經浮現在綾香試圖用音韻地位描述 RIME 詞典的時候了。

自己又看了下之前試圖為韻典網修復的有女羅馬字，結果也有若干脣音小韻在拼寫上的開合有問題。

### 問題二：Qieyun.js 目前默認的音系數據，不符合切韻拼音方案對音系提出的質量標準

這一點是因為：**切韻拼音的完整方案，除了拼音自身以外，還明文提出了對音系的若干質量標準**，這些標準基本上都是要求區分對立指定小韻（或其中部分字）的音韻要素，方案**不推荐在用切韻拼音時使用不符合該質量標準的音系**。這就導致了目前的推導代碼要靠判斷「字頭」或「個別音韻地位」來儘量使音系達到質量標準（不提供除咍韻脣音以外的任何選項也同樣原因）。同時這也是韻鑒 app 的音系資料要採用 @untunt 的 fork 版本，以及 @syimyuzya 也要 fork RIME 切韻拼音的原因。

## 怎樣解決？

第一個問題的要點是：Qieyun.js 目前版本跟它默認的音韻資料仍然有深耦合，具體表現為：

- `音韻地位.檢查` 的判斷標準僅能適配自帶的資料
- `音韻地位.條目`、`.反切`、`.代表字` 都是查詢（query）操作，但 API 卻是「屬性」形式（即僅能調用 getter 得到固定結果），於是音韻地位就和 Qieyun.js「默認的資料集」綁定了
  - 雖然 @ayaka14732 提到過要提供「聯合查詢」，但無論如何，聯合查詢也是資料「查詢」，調用方式顯然不應像 getter 這樣死板

第二個問題則是需要照顧音韻地位處理邏輯的簡潔性，如果直接就往資料裡加入各種鮮見對立，會更多地干擾到常規的處理，讓比如推導方案等變得不好寫。其解決方法則是由我們直接先提供好相關的用於簡化邏輯的 API，推導方案直接使用即可獲得更規整的地位。

於是大體的解決策略就是：

- 音韻地位與資料進一步拆分，解綁後音韻地位作為資料的 key 而存在，自身僅處理六要素直接相關的操作（如 `.屬於` 等）
- 音韻地位自身的 `.檢查` 標準就也可以放寬，需要嚴的時候再嚴（具體見下述）
- 淘汰（deprecate）音韻地位的 `.條目` 等查詢操作以及 `iter音韻地位` 等，相關操作移至數據模塊

## 改進目標

- API 方面：使音韻地位支持無歧義描述切韻拼音、韻典、poem 的音韻地位
- 數據方面：默認數據升級以達到切韻拼音所要求的標準

## API設計概要

（目前為概要，代碼也只是大概樣子）

### 音韻地位與數據分離

```javascript
import Qieyun, { 音韻地位 } from 'qieyun';

// 自帶的一套（未來可能擴成多套）音韻資料置於 `Qieyun.data`
const data = Qieyun.data.defaultData;

// 舊
Qieyun.iter音韻地位();
// 新
data.iter音韻地位();
data.iter小韻();
data.iter條目();
//Qieyun.data.iter音韻地位(); // 多資料聯合 iter，可能要比單資料複雜一些，之後再設計

const 地位 = 音韻地位.from描述(...)；

// 舊
地位.代表字;
地位.反切;
// 新
data.query小韻(地位).map(x => [x.代表字, x.反切]);
// 多套資料「聯合搜索」可能比查詢單套資料更複雜，可以之後再設計

// 不變
地位.屬於`...`
```

### 音韻地位檢查放寬

```javascript
import Qieyun, { 音韻地位 } from 'qieyun';

import { tshet, blankego } from 'qieyun-examples';

// 舊版會因「呼」和「重紐」不合法而立即 throw，新版不會
const 地位 = 音韻地位.from描述('幫合清入');

// 但也不用無限制放寬，放寬到目前夠用即可（之後如果有人請求再放寬的話到時再說）
const 地位2 = 音韻地位.from描述('幫合冬入'); // throw 冬韻 is 呼-neutral, 呼 should be null
const 地位2 = 音韻地位.from描述('幫合之平'); // throw 之韻 is 開口-only, 呼 should be '開'

blankego(地位); // => 'pvet' 這是有女全字表對「碧」小韻的處理

// 但是 Qieyun.data 就不用這麼寬容
const data = Qieyun.data.defaultData;
data.query小韻(地位); // 可以直接返回 null（查詢不到）
                      // 或者因為「呼」「重紐」不合法而回報類似 bad request 之類的
data.query小韻(地位, ...); // 未來可以考慮通過一些參數支持近似地位查詢
                           // 至於「辟」「碧」兩小韻和「幫合清入」分別算不算近似，可視具體參數而定
                           // 這個可以之後再設計
```

需要立即放寬的限制只有兩個：

- 分開合的韻，脣音允許「呼」不為 `null`
- 陽韻脣音、蒸尤幽韻牙喉音允許「重紐」不為 `null`

### 推導方案創建與聲明

```javascript
// 提供推導代碼的包裝函數（即類似 qieyun-example 中的那個 wrapper）

const tshet = Qieyun.createDeriver(code, ...); // 其他參數待定

// 向推導方案查詢選項時，可以讓推導方案回報更多信息：

// blankego.js
if (!音韻地位) {
  return {
    選項: [...], // 和之前格式一致
    音韻地位features: {
      脣音開合: {
        呼: true, // 聲明本代碼可以處理帶「呼」的脣音地位
        fill呼: '廣韻', // 聲明若傳入的呼為 null，則傳入的地位會用類似《廣韻》的方式填入該韻默認的脣音呼再傳給本代碼
      },
      ...
    },
  };
}
// END blankego.js

// legacy-default-features.js
if (!音韻地位) {
  return [...]; // 「兼容模式」：若返回 Array，則模仿 <= 0.12 時代資料的音韻地位features
                // （保質期過後可以棄用「兼容模式」，此代碼將與下一段等同）
}
// END legacy-default-features.js

// default-features.js
if (!音韻地位) {
  return {
    選項: [...]; // 使用新版的默認features（一段時期後將允許簡寫為返回 Array）
  };
}
// END default-features.js

// tshet.js
if (!音韻地位) {
  return {
    選項: [...],
    音韻地位features: {
      脣音開合: {
        咍: 'optional', // 預留值，目前階段完全等同於 true（可處理）
                        // 區別僅在於等未來推導器支持讓用戶選擇不同的 字→音韻地位 這一步的處理選項時，
                        // 'optional' 會指示默認不開啟此項
      },
      蒸幽重紐: 'recommended', // 預留值，目前階段等同於 true
      端知類隔: { // 聲明是否支持處理端組二三等地位，默認是均不支持（只有「地」必須處理不可選）
         default: 'disrecommended', // 預留值，目前階段等同於 false（見下）
         庚二麻三: 'recommended',
      },
      嚴凡對立: false, // false 表示雖然可以傳入嚴韻脣音或凡韻非脣音，但傳入的這種地位會被自動改成規整的
                       // 代碼內只需處理 normalize 後的
      ...
    },
  };
}
// END tshet.js
```

注意這個的主要作用是聲明該代碼支持處理什麼，不支持處理什麼，現階段不能（以後也不完全能）代替選項。

這個 `音韻地位features` 詳見下一節。

#### 為什麼要改韻書資料？

不是改韻書，是目前 `qieyun-data` 的默認音韻數據**質量達不到**現在提出的**標準**。新的默認資料將會是以 poem 表為基礎引入必要的對立而成。

#### 之前說了韻書資料裡的地位要改成不靠人工賦與地位，而是靠反切等數據客觀地處理出來的吧？

文獻工作是不可能避免的。

對個別條目做個別修訂也是不可能避免的。

#### 那只是上游吧？各負責各的就行

不是。

音韻學是都要一塊幹的。

#### 所以要改？

改。

不過會先只改最必要的（即：切韻拼音文章明文提及過的對立），解決當務之急。除此之別的校訂，比如韻鑒還引入了 unt 一些額外修訂等，都可以都先保守些，比如至少可以等到新版韻鑒發佈後再說。

此外最好能維一個調整記錄。

#### 推導方案怎麼辦？

之前說①讓推導方案給出「音韻地位features」②音韻地位與數據分離，都是為這個準備的。

### 修正目標

- 默認資料稱為 `Qieyun.data.defaultData`，暫時（暫時！）獨立於所有韻書資料
  - 等到 nk2028 設想的「資料聯合查詢 API」成熟後再考慮用它去取代這個 `defaultData`
- 原本的 `qieyun-data` 下的廣韻表（來自 poem 字音表）作為原始資料，保存並暫時不再動。
- 默認資料會基於原表並加入以下五項對立：
  - 庚麻韻端組（「爹」「打」）
  - 尤幽蒸韻重紐（「尤B」即是指《廣韻》劃入尤韻的幽B）
  - 「倄」云母
  - 「𩦠」小韻的「並三A陽上」
  - 脣音咍韻

### `音韻地位features`

供推導方案使用，格式為嵌套對象。

取值（下稱 `Option` 型）的含義：

- 表示代碼不處理該類音韻地位：
  - `'reject'`：預留值，目前不用（打算未來用於表示不支持處理，傳入該類地位將直接報錯）
  - `false`：代碼自身不處理該音韻地位，但允許傳入，傳入後會先被轉換成更規整的地位
- 表示代碼可處理該類地位：
  - `true`, ：代碼能夠處理該類音韻地位
  - `'disrecommended'`, `'optional'`：預留值，目前視同 `false`（未來用於表示代碼能夠處理，但指示默認情況下應當不使用該類地位，`'disrecommended'` 額外指示用戶界面應當明顯提示用戶「最好不啟用」）
  - `'recommended'`：預留值，目前視同 `true`（未來用於提示默認情況下應當使用該類地位，`'recommended'` 額外指示用戶界面應當明顯提示用戶「最好啟用」）

目前計劃支持的 features 如下：

（括號內為默認值，選項默認為 `Option` 型，「兼容模式」為上文所述為 <=0.12 代碼準備的默認值）：

- `脣音開合`：可處理脣音地位開合相關信息
  - `咍 (false)`：可處理幫組咍韻
  - `呼 ('reject')`：可處理「呼」不為 `null` 的脣音地位（若為 `false` 則指傳入後會移除脣音地位中的「呼」）
  - `require呼: bool (false)`：僅當可處理「呼」時生效，表示不可處理「呼」為 `null`（除非開合中立韻）的地位
  - `fill呼?: '廣韻' | '切韻' (null)`：僅當可處理呼且不 `require呼` 時生效，當傳入的脣音呼為 `null` 且非開合中立韻時為其填入一個值：`'廣韻'` 表示歌、寒韻脣音填入合口，`'切韻'` 表示填入開口
- `端知類隔`：可處理端知組與等不搭配的情型（注意「地」必須要能處理，不受feature影響）
  - `庚二麻三 (true, 兼容模式 false)`：「打」「爹」等小韻（也包括後起小韻「嗲」等）
  - `端組二三等 (false)`：可處理其他端組二三等地位
  - `知組一四等 (false)`：可處理知組一四等地位
- `'云母非三等'`：
  - `倄 (true, 兼容模式 false)`：可處理倄小韻（云灰上）
  - `其他 ('reject')`：可處理其他云母非三等小韻
- `蒸幽重紐 (false)`：可處理蒸幽韻牙喉音B類
- `脣音陽A (false)`：可處理「𩦠」小韻的「並三A陽上」地位
- `嚴凡對立 (false)`：可處理嚴韻脣音或凡韻非脣音

代碼中可以使用一些簡寫：

```javascript
{
  端知類隔: true, // 表示該組內所有 Option 型值均設為該值
}

{
  端知類隔: {
    庚二麻三: 'recommended',
    default: 'disrecommended', // 表示該選項組所有其他 Option 型值
  }
}

{
  preset: 'poem',  // 引入已知的 preset 然後再覆蓋個別項
                   // 此處 'poem' 表示是「兼容模式」默認值
}
```

（早期開發階段可以先僅支持指定 preset，之後再逐漸支持各選項）

### 實現本文的所有改進案後，切韻拼音推導代碼會有哪些簡化？

```javascript
// 舊：判斷個別聲韻
if (音韻地位.描述 === '影開三蒸入' && '抑𡊁𢬃𡊶'.includes(字頭) ||
  音韻地位.描述 === '溪開三蒸平') {
  return '冰'; // 抑、硱小韻歸 B 類
}
if (音韻地位.描述 === '溪三尤平' && '𠁫𰀧㲃恘惆𣪘'.includes(字頭)) {
  return '幽'; // 𠁫小韻歸幽韻
}

// 新：通過 features 允許推導器直接傳入「蒸B」「尤B」地位
return 音韻地位.判斷([
  ['蒸韻 (幫組 或 合口 或 重紐B類)', '冰'],
  ...,
  ['尤韻 重紐B類', '幽'],
  ...,
]);


// 舊：判斷個別聲韻
if (音韻地位.描述 === '匣合一灰上' && '倄侑'.includes(字頭)) return 'uojq'; // 倄小韻

// 新：直接允許推導器傳入「云灰上」，後面代碼直接可以處理該地位


// 惟一暫未被改進案覆蓋的，暫需保留的特別判定
if (音韻地位.描述 === '云開三眞平') return 'ghin'; // 礥小韻
```

### RIME 中古音上游碼表使用的音韻地位

脣音可以默認「呼」為 null，只有 kyonh 表中的嚴拼和 `fill呼 = '廣韻'` 規則不一致時才需要標出（歌開、寒開等）

**Comment by syimyuzya** [comment]
補充一下 `音韻地位features` 的預想使用效果：

```js
function oldSchema(音韻地位, 字頭, 選項) {
  if (!音韻地位) return [];
  return 音韻地位.描述;
}
function newSchema(音韻地位, 字頭, 選項) {
  if (!音韻地位) return { 音韻地位features: { 脣音開合: { 呼: true } } };
  return 音韻地位.描述;
}

const oldDeriver = Qieyun.createDeriver(oldSchema);
const newDeriver = Qieyun.createDeriver(newSchema);

const 地位 = Qieyun.音韻地位.from描述('並開一歌上');

oldDeriver(地位); // => 並一歌上
newDeriver(地位); // => 並開一歌上
```

**Comment by untunt** [comment]
也许 `from描述('幫合冬入')` 和 `from描述('幫開冬入')` 两种都兼容更好，因为在很多其他资料里开合是必备字段，现在需要手动 `if 韻 in 常量.開合中立的韻` 才能正常读入。必为开口的韵、必为合口的韵也应同理。此外韵图里确实会遇到它们有（伪）开合对立的情况，这将来可能也是需要录入的（现在𩦠小韵需要明确指定重纽为 A 就已经是这种情况了）

另外还应该增加一个把非标准音韵地位转换为正则音韵地位的功能（它甚至应该用一个推导方案实现！？），类似我之前[这里](https://github.com/nk2028/qieyun-encoder-python/pull/5/commits/a14ddef8b0d6c6945e1dbc576b21d91295cb9ca6)的 `更優音韻地位`（但不要那么严格）

**Reply by syimyuzya** [reply]
好耶～

這樣的話，那個音韻地位正則化（即前文的「features」），可以不限於推導器，而是能獨立使用，可以在任何場合調用。

【以前自己想過類似做法來着，但那時顧慮到會破壞 `.反切` `.代表字` `.條目` 就沒做。現在這三樣直接取消了，如何執行「近似查詢」完全交給資料模塊裁定，就沒有那個阻礙了ww

**Reply by ayaka14732** [reply]
> 也许 from描述('幫合冬入') 和 from描述('幫開冬入') 两种都兼容更好，因为在很多其他资料里开合是必备字段，现在需要手动 if 韻 in 常量.開合中立的韻 才能正常读入。必为开口的韵、必为合口的韵也应同理。此外韵图里确实会遇到它们有（伪）开合对立的情况，这将来可能也是需要录入的（现在𩦠小韵需要明确指定重纽为 A 就已经是这种情况了）

我也是這樣想的

> 另外还应该增加一个把非标准音韵地位转换为正则音韵地位的功能

我也是這樣想的

> （它甚至应该用一个推导方案实现！？）

要整合到 library 中

**Comment by syimyuzya** [comment]
按 @untunt 所說的，「呼」也允許在任何韻下都不為 `null`（當然，不可以在需要開合的時候反而為 `null`）。

對於開合中立的韻，它的處理法可以和脣音相仿，正則化規則默認為 `false`。

而對於必為開/合的韻，默認的正則化規則可以設定為 `'reject'`。

現在想到的用法有這些：

### 單獨使用：

默認 preset = `'default'`，即原文所述各項默認值。

若設定 preset 為 `'defaultStrict'`（將 `false` 亦視作 `'reject'`），就與原來的 `音韻地位.驗證` 相仿了。

```js
地位 = 音韻地位.from描述('端合冬平');
地位.正則化(/* 'default' */).描述; // => 端一冬平
地位.正則化('defaultStrict').描述; // Error: 冬韻 is 呼-neutral, 呼 should be null
```

### 構造函數、`from描述` 中指定

默認 preset = `'none'`，即最小程度的限定。

```js
音韻地位.from描述('端合冬平').描述; // => 端合一冬平
音韻地位.from描述('端合冬平', 'default' /* 即 { preset: 'default' } */).描述; // => 端一冬平
```

### 推導器（聲明形式）

默認 preset = `'default'`。對於舊的（返回 `Array` 的）推導代碼，默認 preset = `'poem'`。

```js
if (!音韻地位) {
  return { 音韻地位features: ... };
}
```

### 推導器（代碼中調用）

```js
if (!音韻地位) return { 選項: [], 音韻地位features: 'default' };
const 原地位 = 音韻地位;
音韻地位 = 音韻地位.正則化('poem');
```

#### 有直接調用的形式，為什麼還要保留聲明形式？

- 新推導代碼反正也要用 object 形式返回選項列表（一段時期內會和返回 Array 的舊代碼相區別），不如讓它多幹點事
- 增加的工作量並不大

【當然，如果大家都覺得第一條（使 0.12 時代的推導代碼一段時間內不用改仍能直接用）並不重要，那就可以取消😆】

**Reply by ayaka14732** [reply]
沒有看懂

<img width="180px" src="https://user-images.githubusercontent.com/68557794/160269154-b6829e9e-e3b3-49ca-866e-ee70cfefca24.png">

**Reply by syimyuzya** [reply]
就是允許讓推導方案對推導器說比如：

- 推導方案代碼：我不能推導（也不想看到）『端組庚韻』這種東西，請直接把那些音當成知組給我
- `Qieyun.createDeriver`：好，我給你包裝一層，之後所有傳入的這類地位都會改了再傳給你

再比如：

- 推導方案代碼：我比上面那方案區分更細，我能推導蒸幽韻的重紐，給我傳 `音韻地位` 時不要把它們的 B 類丟掉
- `Qieyun.createDeriver`：好，我給你包裝的層裡會把它們原樣給你

【哎……這玩意要是 Python 的話就是個 decorator 了 233

**Reply by syimyuzya** [reply]
至於**為什麼要有這個機制**……因為要是沒有這個，就相於讓所有推導方案都必須照顧到數據裡所有的對立。

比如有了這個，推潘悟云擬音就可以：

```js
// 聲明了本代碼不接受蒸幽重紐後
if (is`重紐B類`) 介 = 'ɯ' + 介;
```

否則就得要麼：

```js
// 所有之前的代碼都要在 `if (!音韻地位) return ...` 和 `const is = ...` 之間加上這句
音韻地位 = 音韻地位.正則化(...);
```

要麼：

```js
// 潘的蒸、幽韻用 ɨ，想給它加重紐也加不了
if (is`重紐B類 非 蒸幽韻`) 介 = 'ɯ' + 介;
// 讓代碼平白無故增加了無關內容，而且還在邏輯表達式裡，會很影響可讀性
// qieyun-examples 既有代碼也都要大改才行了
```

（當然，如果覺得不需要照顧兼容性，直接要求過去寫的推導代碼想在新版用就要加上 `音韻地位 = 音韻地位.正則化();`，也是不錯的解決方式ww

**Comment by syimyuzya** [comment]
這個「正則化」還有一種設計想法，即：不作為對象方法，而是作為函數。

效果對比：

```js
// 作為方法
地位.正則化('poem');
// 作為函數
音韻地位.正則化(地位, 'poem');
// 柯里化風格（Node.js 傳 統 藝 能 www
音韻地位.正則化('poem')(地位);

// 先後適用兩個正則化規則
// 作為方法
地位.正則化(config1).正則化(config2);
// 作為函數
音韻地位.正則化(音韻地位.正則化(地位, config2), config1);
// 柯里化風格
音韻地位.正則化(config2)(音韻地位.正則化(config1)(地位));

// 與自定義函數相組合
function 自定義變換(地位) { ... }
// 方法
自定義變換(地位.正則化(config1)).正則化(config2);
// 函數
音韻地位.正則化(自定義變換(音韻地位.正則化(地位, config1)), config2);
// 柯里化風格
音韻地位.正則化(config2)(自定義變換(音韻地位.正則化(config1)(地位)));
```

在推導器裡，由於 `音韻地位` 這名字已經被佔了，要寫成 `Qieyun.音韻地位.正則化(音韻地位)` 或 ~~`音韻地位.prototype.正則化(音韻地位)`~~ `音韻地位.constructor.正則化(音韻地位)` 之類的。

---

PS 之前甚至還想過要不要把它拆成若干小函數，完全靠組合來使用……然而不行，因為用「選項」能做到「基於某個 preset 但是除去其中某項」，而函數組合很難去除其中某項。

像這樣：

```js
// 用選項
地位.正則化({ preset: 'default', 幫組開合: { 咍韻: true }}); // 用 default 但是不要把幫組咍韻轉成灰韻
// 用函數組合
??? // 寫不出來，只能重新從小函數開始組合
```

**Reply by ayaka14732** [reply]
> `音韻地位` 這名字已經被佔了

推導器裏的以後可以改稱 `地位`

**Reply by syimyuzya** [reply]
> > `音韻地位` 這名字已經被佔了
> 
> 推導器裏的以後可以改稱 `地位`

😂😂

（不過話說現在 `class 音韻地位` 本來就不在代碼的 namespace 裡（以後也不用加入），代碼真想頻繁用它的話無論如何也得自行來一句 `const { 正則化 } = Qieyun.音韻地位;` 什麼的，並不會更簡省（（

**Reply by graphemecluster** [reply]
我覺得同時實現兩種也無妨吧
```ts
function 自定義變換(地位) { … }
const 變換 = 音韻地位.產生正則變換(config1, 自定義變換);
地位.正則化(config2, 自定義變換, 變換).正則化(config3);
```

大概是這樣實現：
```ts
type Transformer = (地位: 音韻地位) => 音韻地位;
type Config = ...;
static 產生正則變換(...configs: (Config | Transformer)[]): Transformer {
	function generateTransformer(config: Config): Transformer { ... }
	return 地位 => configs.reduce((地位, config) => (typeof config === "function" ? config : generateTransformer(config))(地位), 地位);
}
正則化(...configs: (Config | Transformer)[]): 音韻地位 {
	return 音韻地位.產生正則變換(...configs)(this);
}
```

**Reply by syimyuzya** [reply]
> 我覺得同時實現兩種也無妨吧
> 
> ```ts
> function 自定義變換(地位) { … }
> const 變換 = 音韻地位.產生正則變換(config1, 自定義變換);
> 地位.正則化(config2, 自定義變換, 變換).正則化(config3);
> ```
> 
> 大概是這樣實現：
> 
> ```ts
> type Transformer = (地位: 音韻地位) => 音韻地位;
> type Config = ...;
> static 產生正則變換(...configs: (Config | Transformer)[]): Transformer {
> 	function generateTransformer(config: Config): Transformer { ... }
> 	return 地位 => configs.reduce((地位, config) => (typeof config === "function" ? config : generateTransformer(config))(地位), 地位);
> }
> 正則化(...configs: (Config | Transformer)[]): 音韻地位 {
> 	return 音韻地位.產生正則變換(...configs)(this);
> }
> ```

Pipe 形式好耶w
